src/datasources/prices-api/prices-api.service.ts

```diff
@@ -0,0 +1,118 @@
+      this.configurationService.getOrThrow<number>(
+        'expirationTimeInSeconds.default',
+      );
+    this.pricesTtlSeconds = this.configurationService.getOrThrow<number>(
+      'prices.pricesTtlSeconds',
+    );
+    this.defaultNotFoundExpirationTimeSeconds =
+      this.configurationService.getOrThrow<number>(
+        'expirationTimeInSeconds.notFound.default',
+      );
+  }
+
+  async getNativeCoinPrice(args: {
+    nativeCoinId: string;
+    fiatCode: string;
+  }): Promise<AssetPrice> {
+    try {
+      const cacheDir = CacheRouter.getNativeCoinPriceCacheDir(args);
+      const url = `${this.baseUrl}/simple/price?ids=${args.nativeCoinId}&vs_currencies=${args.fiatCode}`;
+      return await this.dataSource.get({
+        cacheDir,
+        url,
+        networkRequest: this._buildNetworkRequest(),
+        notFoundExpireTimeSeconds: this.defaultNotFoundExpirationTimeSeconds,
+        expireTimeSeconds: this.pricesTtlSeconds,
+      });
+    } catch (error) {
+      throw new DataSourceError(
+        `Error getting ${
+          args.nativeCoinId
+        } price from provider${this._mapProviderError(error)}`,
+      );
+    }
+  }
+  async getTokenPrice(args: {
+    chainName: string;
+    tokenAddress: string;
+    fiatCode: string;
+  }): Promise<AssetPrice> {
+    try {
+      const cacheDir = CacheRouter.getTokenPriceCacheDir(args);
+      const url = `${this.baseUrl}/simple/token_price/${args.chainName}?contract_addresses=${args.tokenAddress}&vs_currencies=${args.fiatCode}`;
+      const result: AssetPrice = await this.dataSource.get({
+        cacheDir,
+        url,
+        networkRequest: this._buildNetworkRequest(),
+        notFoundExpireTimeSeconds: this.defaultNotFoundExpirationTimeSeconds,
+        expireTimeSeconds: this.pricesTtlSeconds,
+      });
+      return result;
+    } catch (error) {
+      throw new DataSourceError(
+        `Error getting ${
+          args.tokenAddress
+        } price from provider${this._mapProviderError(error)}`,
+      );
+    }
+  }
+
+  async getFiatCodes(): Promise<string[]> {
+    try {
+      const cacheDir = CacheRouter.getPriceFiatCodesCacheDir();
+      const url = `${this.baseUrl}/simple/supported_vs_currencies`;
+      const result: string[] = await this.dataSource.get({
+        cacheDir,
+        url,
+        networkRequest: this._buildNetworkRequest(),
+        notFoundExpireTimeSeconds: this.defaultNotFoundExpirationTimeSeconds,
+        expireTimeSeconds: this.defaultExpirationTimeInSeconds,
+      });
+      return result;
+    } catch (error) {
+      throw new DataSourceError(
+        `Error getting Fiat Codes from prices provider${this._mapProviderError(
+          error,
+        )}`,
+      );
+    }
+  }
+
+  private _mapProviderError(error: any): string {
+    const errorCode = error?.status?.error_code;
+    return errorCode ? ` [status: ${errorCode}]` : '';
+  }
+
+  private _buildNetworkRequest(): NetworkRequest | undefined {
```

Comment:
This function only sets the authentication headers when the `apiKey` is present (and therefore a paid subscription is being used). 

I was wondering if we could include a `timeout` parameter also into these Coingecko calls, just to prevent our backend from waiting too much for the response. Waiting for a price for a long time seems unacceptable, what do you think @fmrsabino?

Reply:
We should set a timeout for these requests. For now I'd use `httpClient.requestTimeout` from the configuration.